egs/wsj/s5/steps/dur_model/train.sh

```diff
@@ -0,0 +1,183 @@
+    exit 1
+  fi
+else
+  parallel_train_opts="--use-gpu=no"
+fi
+
+if [ $stage -le 0 ]; then  # do the nnet-related initializations only if stage<=0
+  echo "$0: Initializing nnet3 model..."
+  stage=0
+  if [[ ! -z $nnet_config ]]; then
+    echo "$0: Using provided config file for nnet: $nnet_config"
+  else
+    nnet_config=$dir/nnet.conf
+    feat_dim=$(durmodel-info $durmodel | grep feature-dim | awk '{ print $2 }') || exit 1;
+    output_dim=$(durmodel-info $durmodel | grep max-duration | awk '{ print $2 }') || exit 1;
+    steps/dur_model/make_nnet_config.sh $nnet_opts $feat_dim $output_dim >$nnet_config
+    echo "$0: Wrote nnet config to "$nnet_config
+  fi
+  $cmd $dir/log/nnet_init.log \
+       nnet3-init $nnet_config $dir/nnet.raw || exit 1;
+  $cmd $dir/log/nnet_durmodel_init.log \
+       nnet3-durmodel-init $durmodel $dir/nnet.raw $dir/0.mdl || exit 1;
+else
+  [ ! -f $dir/$stage.mdl ] && echo "$0: Nnet-duration model file not found (you provided --stage=$stage): $dir/$stage.mdl" && exit 1;
+fi
+
+num_examples=`cat $dir/num_examples` || num_examples=1
+num_epochs_per_iter=$[40000000/$num_examples]
+[[ $num_epochs_per_iter -le 0 ]] && num_epochs_per_iter=1
+[[ $num_epochs_per_iter -gt 8 ]] && num_epochs_per_iter=8
+num_iterations=$[$num_epochs/$num_epochs_per_iter]
+
+echo "$0: Number of epochs per each iteration is $num_epochs_per_iter"
+echo "$0: Will train from iteration $stage through iteration $[$num_iterations-1] ..."
+
+for iter in $(seq $stage $[$num_iterations-1]); do
+  echo "Iteration: "$iter
+  curr_mdl=$dir/$iter.mdl
+  next_mdl=$dir/$[$iter+1].mdl
+
+  train_egs="ark:for n in $(seq -s ' ' $num_epochs_per_iter); do cat $dir/train.egs; done |"
+  $cmd $train_queue_opt $dir/log/train.$[$iter+1].log \
+       nnet3-train $parallel_train_opts "nnet3-durmodel-copy --raw=true $curr_mdl -|" \
+            "ark:nnet3-shuffle-egs --buffer-size=$shuffle_buffer_size --srand=$iter \
+            '$train_egs' ark:-| \
+            nnet3-merge-egs --minibatch-size=$minibatch_size ark:- ark:- |" \
+            $dir/nnet.raw || exit 1;
+
+  $cmd $dir/log/durmod_set_raw_nnet.log \
+       nnet3-durmodel-copy --set-raw-nnet=$dir/nnet.raw $curr_mdl $next_mdl
+
+  $cmd $dir/log/compute_prob_valid.$[$iter+1].log \
+     nnet3-compute-prob $dir/nnet.raw \
+     "ark:nnet3-merge-egs --minibatch-size=$minibatch_size ark:$dir/val.egs ark:- |" \
+     && grep -o -H "Overall.*" $dir/log/compute_prob_valid.$[$iter+1].log &
+
+  if $early_stop; then
+    wait
+    curr_val_logprob=$(grep -o "Overall log.*" $dir/log/compute_prob_valid.$iter.log 2>/dev/null | awk '{print $6}')
+    next_val_logprob=$(grep -o "Overall log.*" $dir/log/compute_prob_valid.$[$iter+1].log | awk '{print $6}')
+    num_epochs_until_now=$[$iter*$num_epochs_per_iter]
+    if [[ $(echo "$next_val_logprob < $curr_val_logprob" | bc 2>/dev/null) == 1 && $num_epochs_until_now -ge $early_stop_min_epochs ]]; then
```

Comment:
Yes, it looks like that option does not exist in nnet3-compute-prob- it's
straightforward to add though, you can add it-- make it default to "no".
There is some stuff in #ifdefs that some of these programs have, you should
add it.

Regarding the num-epochs-per-iter being configurable-- I think it will be
easier for you, and simpler, to make the num-samples configurable- this
will avoid extra code.  People can change the script if they really need.
Have the script print out how many epochs per iteration it will do.

Dan

On Sun, Jan 24, 2016 at 4:08 PM, Hossein Hadian <notifications@github.com>
wrote:

> In egs/wsj/s5/steps/dur_model/train.sh
> <https://github.com/kaldi-asr/kaldi/pull/414#discussion_r50642372>:
>
> > +            $dir/nnet.raw || exit 1;
> > +
> > +  $cmd $dir/log/durmod_set_raw_nnet.log \
> > +       nnet3-durmodel-copy --set-raw-nnet=$dir/nnet.raw $curr_mdl $next_mdl
> > +
> > +  $cmd $dir/log/compute_prob_valid.$[$iter+1].log \
> > +     nnet3-compute-prob $dir/nnet.raw \
> > +     "ark:nnet3-merge-egs --minibatch-size=$minibatch_size ark:$dir/val.egs ark:- |" \
> > +     && grep -o -H "Overall.*" $dir/log/compute_prob_valid.$[$iter+1].log &
> > +
> > +  if $early_stop; then
> > +    wait
> > +    curr_val_logprob=$(grep -o "Overall log.*" $dir/log/compute_prob_valid.$iter.log 2>/dev/null | awk '{print $6}')
> > +    next_val_logprob=$(grep -o "Overall log.*" $dir/log/compute_prob_valid.$[$iter+1].log | awk '{print $6}')
> > +    num_epochs_until_now=$[$iter*$num_epochs_per_iter]
> > +    if [[ $(echo "$next_val_logprob < $curr_val_logprob" | bc 2>/dev/null) == 1 && $num_epochs_until_now -ge $early_stop_min_epochs ]]; then
>
> So I will put them in the same GPU job. However nnet3-compute-prob does
> not support using a GPU (according to the code). Is that right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/414/files#r50642372>.
>